src/storage/mod.rs

```diff
@@ -1223,15 +1243,18 @@ impl<E: Engine> Storage<E> {
         Err(Error::InvalidCf(cf.to_owned()))
     }
 
-    fn check_key_ranges(ranges: &[KeyRange]) -> bool {
+    fn check_key_ranges(ranges: &[KeyRange], reverse: bool) -> bool {
         let ranges_len = ranges.len();
         for i in 0..ranges_len {
             let start_key = ranges[i].get_start_key();
             let mut end_key = ranges[i].get_end_key();
             if end_key.is_empty() && i + 1 != ranges_len {
                 end_key = ranges[i + 1].get_start_key();
             }
-            if !end_key.is_empty() && start_key >= end_key {
+            if !end_key.is_empty()
+                || !reverse && start_key >= end_key
+                || reverse && start_key <= end_key
```

Comment:
It seems like the ranges are invalid when the end_key is not empty 🤔 
May be you want to write
```rust
if !end_key.is_empty()
    && (!reverse && start_key >= end_key)
    && (reverse && start_key <= end_key)
```

In fact, I suppose that when the `KeyRange` is used, the range it represents should be [start_key, end_key). I means, you don't need to change the function `check_key_ranges`.

Reply:
oh. I made a mistake.